### PR TITLE
AzureMonitor: display errors from requests for the dropdowns

### LINKED ISSUE #31109: AzureMonitor: Add error handling to Metrics
Related to https://github.com/grafana/grafana/issues/30794

### PR DESCRIPTION
**Pull request by joshhunt**
**What this PR does / why we need it**:

The Azure Monitor query editor makes a number of frontend request to populate the dropdowns used to form the query, and any of these requests can fail (for whatever reason, but most often probably due to incorrectly using template variables?), and we should show this to the user.

It's important to note that this is not executing the data source query itself (that's done in the backend), but getting metadata for the UI.

This PR adds frontend error handling for all the frontend calls that the plugin currently makes. Each child dropdown which requests data will call `setError(errorSource: string, error: Error)` where the parent will collect all errors in an tuple-array (so we can keep track of the most recent error) with the source, and the next time that component successfully requests data, it clears the error out with `setError(errorSource, undefined)`.

The top-level QueryEditor component then will show only the most recently reported error in the UI (after running the error through a scary function to extract the error message from wherever it is in the Error-like object). 

Note that this exposes an issue with how the fields are "unset" by setting them to the string `"select"`, and those values are sent to the Azure API. In a future PR i will fix this by setting them to undefined instead.

**Which issue(s) this PR fixes**:

Fixes #31109

**Special notes for your reviewer**:

I know this might seem a little overcomplicated. I would have preferred to actually just clear out the error once the backend datasource sucessfully executes, but i couldnt manage a way to do that. This though is probably the more 'correct' approach, as the errors are cleared as soon as they're no longer applicable.

### CODE REVIEW

**Review by sunker — APPROVED** [review]
Not tested this, but I like the overall approach!

File: public/app/plugins/datasource/grafana-azure-monitor-datasource/components/QueryEditor/messageFromError.ts [hunk 1]
```diff
@@ -0,0 +1,33 @@
+export function messageFromError(error: any): string | undefined {
```
[the remark this answers is not in the document]
Comment (joshhunt) [reply public/app/plugins/datasource/grafana-azure-monitor-datasource/components/QueryEditor/messageFromError.ts:1]: wait really? 🤢 lol

File: public/app/plugins/datasource/grafana-azure-monitor-datasource/components/MetricsQueryEditor/MetricsQueryEditor.test.tsx [hunk 2]
```diff
@@ -22,6 +22,7 @@ describe('Azure Monitor QueryEditor', () => {
         datasource={mockDatasource}
         variableOptionGroup={variableOptionGroup}
         onChange={() => {}}
+        setError={() => {}}
```
Comment (dsotirakis) [inline public/app/plugins/datasource/grafana-azure-monitor-datasource/components/MetricsQueryEditor/MetricsQueryEditor.test.tsx:25]: Could we also have some tests here? Or do you plan to have them elsewhere?

Comment (joshhunt) [reply public/app/plugins/datasource/grafana-azure-monitor-datasource/components/MetricsQueryEditor/MetricsQueryEditor.test.tsx:25]: Yeah, tests were coming but wanted to get more feedback on my approach. seems good to continue with it for now

**Review by mckn — COMMENTED** [review]
I think this looks really good! I am guessing that you don't want to use an `ErrorBoundary` because you still want to show the editor UI even if you get an error?

**Comment by joshhunt** [comment]
@mckn ErrorBoundary will catch exceptions in the render function, right? Here's im trying to catch errors that come from API requests which wouldnt get caught by the ErrorBoundary

**Review by sunker — APPROVED** [review]

**Comment by mckn** [comment]
> @mckn ErrorBoundary will catch exceptions in the render function, right? Here's im trying to catch errors that come from API requests which wouldnt get caught by the ErrorBoundary

Ah, true!

**Pull request merged**